versions/3.0.4.md

```diff
@@ -133,10 +133,22 @@ In order to preserve the ability to round-trip between YAML and JSON formats, YA
 
 ### <a name="documentStructure"></a>OpenAPI Description Structure
 
-An OpenAPI Description MAY be made up of a single document or be divided into multiple, connected parts at the discretion of the author. In the latter case, `$ref` fields MUST be used in the specification to reference those parts as follows from the [JSON Schema](https://json-schema.org) definitions.  In a multi-document description, the document containing the [OpenAPI Object](#oasObject) is known as the **entry OpenAPI document.**
+An OpenAPI Description (OAD) MAY be made up of a single document or be divided into multiple, connected parts at the discretion of the author. In the latter case, [Reference Object](#referenceObject) and [Path Item Object](#pathItemObject) `$ref` keywords, as well as the [Link Object](#linkObject) `operationRef` keyword, are used.  In a multi-document description, the document containing the [OpenAPI Object](#oasObject) is known as the **entry OpenAPI document.**
 
 It is RECOMMENDED that the entry OpenAPI document be named: `openapi.json` or `openapi.yaml`.
 
+#### <a name="structuralInteroperability"></a>Structural Interoperability
+
+When parsing an OAD, JSON or YAML objects are parsed into specific Objects (such as [Operation Objects](#operationObject), [Response Objects](#responseObject), [Reference Objects](#referenceObject), etc.) based on the parsing context.  Depending on how references are arranged, a given JSON or YAML object can be parsed in multiple different contexts:
+
+* As a full OpenAPI Description document (an [OpenAPI Object](#oasObject) taking up an entire document)
+* As the Object type implied by its parent Object within the document
+* As a reference target, with the Object type matching the reference source's context
+
+In version 2.0 of this specification, this section specified that the OAD is a "single file", which (together with the treatment of `$ref` as a URL resolved separately each time it is encountered), resulted in the common interpretation that if the same object is parsed multiple times in contexts that require it to be parsed as _different_ Object types, then as long as the reference target is valid for each Object type, then there is no error.  An example would be parsing an empty JSON object as a reference target twice: Once as a Path Item Object and once as a Schema Object.  Since both Object types allow empty objects, this is syntactically valid for both Objects.
+
+This approach to conflicting contexts remains valid in version 3.0, although it is not required by this version's text.  However, it is RECOMMENDED that description authors avoid scenarios involving conflicting contexts to maximize the portability of their OADs across tools, as well as to version 3.1 where the result of such conflicting contexts is _implementation-defined_ and MAY be treated as an error.
```

Comment:
@DavidBiesack thanks- I agree with the improvement, but will probably wait until I can get some confirmation about whether that line is even correct before updating it.  The strange "this version's text" wording is me flailing for some way to say _"I've been told that this is what is supposed to happen by people who are either no longer involved or not currently responding with details, but I can't find anywhere in the text that justifies that position, which appears to be carried over from 2.0, despite the fact that the text in 2.0 that made that mostly clear has definitely been removed, which I would personally interpret as changing the requirements, but to what I am not sure."_

That seemed over-long, though 😅 (not to mention too much of a negative rant – it's important to me to respect past work whether I currently understand it or not.  The folks who made that decision had a reason, I just don't know what it was).  So yeah, "not required by this version's text" is my weirdly legalistic way of saying ¯\\_(ツ)_/¯ and hoping someone will see it and explain.

Reply:
Regarding "although it is not required by this version of the OpenAPI Specification":

My impression is that it is explicitly *strongly suggested* by v3.0.x:

> **[4.7.23 Reference Object](https://spec.openapis.org/oas/v3.0.3#reference-object)**
> [...]
> The Reference Object is defined by [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03) and follows the same structure, behavior and rules.
> 
> For this specification, reference resolution is accomplished as defined by the JSON Reference specification and not by the JSON Schema specification.

And [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-4) says

> Resolution of a JSON Reference object SHOULD yield the referenced
> JSON value.  Implementations MAY choose to replace the reference with
> the referenced value.

It says _SHOULD_ in JSON Reference, so it is not _required_ to do it that way and only _strongly suggested_ 🙄

Reply:
@ralfhandl to be clear, I'm totally open to changing this PR to say "3.0 behaves just like 2.0", I just don't see that as supported by the text.  I'm happy to be overruled, though, as I just want this to be as clear as possible.  Right now I don't find it clear at all.

The part that was in 2.0 that was removed reads:

> The Swagger representation of the API is made of a single file. However, parts of the definitions can be split into separate files, at the discretion of the user.

So that, to me, says "it doesn't matter how anything is arranged, it is the 'dereferenced` single file that matters, and therefore there isn't any context in the target document."  The only context that matters is the reference context, because it is literally just one file in the end.  That has been removed in 3.0, which to me allows for the possibility of context in the target document, or context maintained between parsing.

But again, if the TSC says that 3.0 works the same as 2.0, I'm happy to add that clarification in 3.0.4.

Reply:
> It doesn't say anything about what is done with that JSON value, and whether it makes sense to treat the _same_ referenced value as _different things_.

JSON Reference says

> Implementations MAY choose to replace the reference with
the referenced value.

That this is a "MAY" is even weirder than the "SHOULD" in the preceding sentence 🙄 

I interpret this as "copy&paste" without taking context into account.

Reply:
> 3.0 works the same as 2.0

That's what all my code assumes: 3.0 is a superset of 2.0 with minor structural differences. After ignoring those differences, the building blocks are interpreted the same.

Reply:
MAY means it's completely optional, so (unlike SHOULD) you don't need a reason to ignore it.  So it definitely does not mean that you have to copy-paste.  In fact with JSON Schema draft-04, which also cites that same JSON Reference draft, you _cannot do so_ in all circumstances.  Which is actually true in OpenAPI as well, not that that stops people from trying.

But again, what is the point of debating the JSON Reference wording here?  What we need is for someone to say what the intent was and how it should be written about.  I can find implementations that assume it's the same as 2.0.  I can find implementations that have no clue what 2.0 was like and implement 3.0 differently based on the different wording.

What was the intent?  (probably the same as 2.0 but then why was the wording changed so substantially?) And how should it be described? (I have no clue and don't like what I've written)

Reply:
@ralfhandl in case it's not clear, I largely agree that your interpretation is the most likely intent.  It just doesn't connect enough in my head (given my own reading of things) for me to know what to do here.  To me, this is all in "resulted in the common interpretation" territory.  But that makes it sound like people were interpreting things that aren't there.  And honestly, that's what it looks like to me, but I know that's not what it looks like to others.

But how should it be phrased?  As a SHOULD recommendation?  As implementation-defined?  Is this a MUST that just didn't get spelled out, and if so are we comfortable enough to put it in as a MUST because implementations would have carried it over from 2.0?  I'd be comfortable with that if someone who knows the intent said so.

Reply:
"resulted in the common interpretation", it it's the common interpretation let's record it in the spec text.

> I know that's not what it looks like to others.

Do you have any specific examples of misinterpretation in the context of OpenAPI 3.0.x?

Reply:
All of this language got taken out so I'm marking this resolved.